### PR TITLE
Build the project with GH Actions.

### PR DESCRIPTION
**Pull request by kittaakos**
#### What it does

 - Enables the build on GH Actions. (This PR has no intention of removing the Travis config, renaming the npm scripts, updating the build badge, etc.)
 - Adds `retries` for the Mocha tests. (It does not mean flaky tests will pass from now on, but instead of rerunning the entire test npm script, we rerun the failed test. Do you want to check if a test has failed and the retry fixed it? Go ahead, and find an [appropriate reporter](https://github.com/mochajs/mocha/commit/3670dc4fefc034807564a069683a8bb300542237).)
 - Adds a workaround for the flawed `theia-run.js` logic: https://github.com/eclipse-theia/theia/issues/8779.

What it does **not** do:
 - It does not add the missing `NPM_AUTH_TOKEN` secret to GH. I am not the owner of the repo, I won't make such a change. (By the way, we should also add `ACTIONS_RUNNER_DEBUG` and `ACTIONS_STEP_DEBUG` with `true` value to enable better logging. See the [docs](https://docs.github.com/en/free-pro-team@latest/actions/managing-workflow-runs/enabling-debug-logging).)
 - It does not do anything with the fact that the Electron codec modification happens after the `tsc` build. The details are [here](https://github.com/eclipse-theia/theia/issues/8772#issuecomment-733622141).
 -  I could not verify the npm publishing, although I could from my fork, so if it has a problem, let me know. (I did the same for other projects with GH Actions, and it worked.)
 - It does not cache things for `yarn` and `node_modules` (with `lerna`, but someone had a similar issue, and was solved like [this](https://github.com/markuplint/markuplint/pull/111/files)).

For the _What it does **not** do_ part, I am CCing @marcdumais-work and @benoitf.

#### How to test

 - Check the new config.
 - Check the example builds [here](https://github.com/kittaakos/theia/actions/runs/383444808).

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: .github/workflows/build.yml
Comment (benoitf) [inline .github/workflows/build.yml:22]: I think we should not use aliases there but use real versions
like macos-10.15, ubuntu-18.04 or ubuntu-20.04 as ubuntu-latest is switching to 20.04 soon on github actions

it will prevent to have scripts that are no longer working just due to a switch of runners that are not under control

Comment (kittaakos) [reply .github/workflows/build.yml:22]: > like macos-10.15, ubuntu-18.04 or ubuntu-20.04 as ubuntu-latest is switching to 20.04 soon on github actions

Sure, no problem. Let me know which targets do you prefer. Thanks!

File: .github/workflows/build.yml
Comment (kittaakos) [inline .github/workflows/build.yml:76]: This is indeed a bug 😊

File: .github/workflows/build.yml
Comment (kittaakos) [inline .github/workflows/build.yml:24]: Based on this https://github.com/eclipse-theia/theia/pull/8780/files#r530476212, we can also pin the Python and Node.js version. I do not have a strong opinion on this. For the record: the Travis-based build uses a floating version of Node.js.

File: .github/workflows/build.yml
Comment (benoitf) [reply .github/workflows/build.yml:22]: just replace latest by their current counter part specified now as you already tested these configurations
https://github.com/actions/virtual-environments#available-environments

it's then: `ubuntu-18.04` `macos-10.15` `windows-2019`

File: .github/workflows/build.yml
Comment (kittaakos) [inline .github/workflows/build.yml:7]: TODO

File: .github/workflows/build.yml
Comment (benoitf) [inline .github/workflows/build.yml:36]: may probably require `fetch-depth: "0"`

File: .github/workflows/build.yml
Comment (benoitf) [reply .github/workflows/build.yml:24]: yarn is also automatically provided by the runners so it may have different behaviour over time.
It depends where you want to put the cursor of 'validate automatically against latest' and 'use pre-defined versions'
maybe some jobs could check against latest of everything and the one that publishes should use pre-defined versions

File: .github/workflows/build.yml
Comment (kittaakos) [reply .github/workflows/build.yml:36]: Let's do that. I do not have any objections. Usually, you use `fetch-depth: '0'` if you want to check existing tags/releases in the repo and use it in your step/job condition. Otherwise, the default `1` is faster.

As I started in the PR description, this will be most likely required to make the yarn caching with lerna run properly with GH Actions.

> * It does not cache things for `yarn` and `node_modules` (with `lerna`, but someone had a similar issue, and was solved like [this](https://github.com/markuplint/markuplint/pull/111/files)).

I am going to make the suggested change.

Comment (kittaakos) [reply .github/workflows/build.yml:36]: @benoitf, out of curiosity, did you intentionally mention only this `checkout` action and not the other for the npm publish build? Just asking, though.

Comment (benoitf) [reply .github/workflows/build.yml:35]: @kittaakos in fact yes it was more for the 'build' usecase

File: .github/workflows/build.yml
Comment (marechal-p) [inline .github/workflows/build.yml:23]: Let's use Python 3? https://github.com/eclipse-theia/theia/issues/8383

File: .github/workflows/build.yml
Comment (marechal-p) [inline .github/workflows/build.yml:51]: Bash on Windows too?

File: .github/workflows/build.yml
Comment (marechal-p) [reply .github/workflows/build.yml:79]: That might be for the best because as it is written right now the publish job redoes what the build job does...

File: .github/workflows/build.yml
Comment (marechal-p) [inline .github/workflows/build.yml:81]: So no more `next` publish on each commit from `master`, but every day at 4am?

File: .github/workflows/build.yml
Comment (kittaakos) [reply .github/workflows/build.yml:23]: I would like to postpone it. Travis builds uses Python 2.

File: .github/workflows/build.yml
Comment (kittaakos) [reply .github/workflows/build.yml:51]: Yes.

File: .github/workflows/build.yml
Comment (kittaakos) [reply .github/workflows/build.yml:81]: I wanted to publish when on `master` and **not** a CRON job (e.g., merged PRs.). Where is the bug? I must be blind 😊

File: .github/workflows/build.yml
Comment (kittaakos) [reply .github/workflows/build.yml:79]: > That might be for the best because as it is written right now the publish job redoes what the build job does...


> Could we just split this workflow in 2 ?

This PR has no intention of making the build smarter or more stable than we have now. Feel free to open up a follow-up once this is merged.

File: .github/workflows/build.yml
Comment (marechal-p) [reply .github/workflows/build.yml:81]: Ok I misread the conditional. Can you explain why we want to run the tests at 4am? I did not see a note about this in the PR description.

File: .github/workflows/build.yml
Comment (marechal-p) [inline .github/workflows/build.yml:86]: I overcame this limitation on my PR by doing only one job that conditionally publishes as a last step. No artifact uploading weirdness required.

**Comment by marechal-p** [comment]
>     * It does not do anything with the fact that the Electron codec modification happens after the `tsc` build.

Since we do not do caching anymore the Electron codec scripts are ran on Yarn install, so this is fine.

File: .github/workflows/build.yml
Comment (marechal-p) [reply .github/workflows/build.yml:81]: I understand that this `publish` job runs when we push on master and not when the cron runs, but why the cron at all?

File: .github/workflows/build.yml
Comment (kittaakos) [reply .github/workflows/build.yml:86]: > I overcame this limitation on my PR by doing only one job that conditionally publishes as a last step

I've seen that. I think that would end up in a flawed pipeline as publishing could happen on the Linux node right after the (green) tests step, but the Windows build job could have failed. This PR mimics the current Travis CI behavior: publishing happens only when all builds are green.

To achieve this👆, I usually use multiple jobs with `needs` to create a dependency between the jobs. According to the [`needs context`](https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#needs-context) doc, this will ensure that the previous job has a `result` (`'success'`). So the publish happen only **iff** all builds where green. Do you have a better suggestion?

File: .github/workflows/build.yml
Comment (kittaakos) [reply .github/workflows/build.yml:81]: > but why the cron at all?

I do not know if we have one currently, but we used to have a CRON job from the beginning of this project. I set it up on Travis UI if I remember correctly. [Here](https://travis-ci.org/github/eclipse-theia/theia/builds/691440689) is one build example from 6 months ago.

File: .github/workflows/build.yml
Comment (marechal-p) [reply .github/workflows/build.yml:86]: Right. Thinking about it, to publish all we need is to transpile our TypeScript sources into JavaScript. We could get away just doing `yarn build` then? Except that we need to do `yarn` first and doing this does `install, build, lint, build example apps and download plugins` which is what annoys me the most with our `prepare` script right now. Do you think it'd be better to also just run the build step for the publish?

File: .github/workflows/build.yml
Comment (marechal-p) [reply .github/workflows/build.yml:81]: > I do not know if we have one currently, but we used to have a CRON job from the beginning of this project.

Ok but why? I can't find a reason for it or anything that requires a daily job. If we publish on each commit on master as well as test, then what would change from day to day that needs the CI to run for?

File: .github/workflows/build.yml
Comment (marechal-p) [reply .github/workflows/build.yml:79]: > * one workflow that apply on master branch and build from master branch but running only on linux for example ?

@benoitf this is what I went for in my PR but it turns out that it is more "sure" to test all platforms again on publish:
- The workflow for PRs usually test the incoming branch only.
- The workflow for merged commit will test the latest state for master with the new changes.
- GitHub only performs git conflict checks on PR branches, when merged it can still change the behavior of apps...

Correct me if I am wrong.

**Review by marechal-p — APPROVED** [review]
LGTM, checks are green so CI is back :)

**Comment by vince-fugnitto** [comment]
@kittaakos is something missing before we are able to merge the pull-request, or did you want to perhaps do it after the release?
If something is missing or can be improved we can do it in subsequent prs, having the ability to build and test is already super useful :smiley:

**Comment by kittaakos** [comment]
> @kittaakos is something missing before we are able to merge the pull-request, or did you want to perhaps do it after the release?
> If something is missing or can be improved we can do it in subsequent prs, having the ability to build and test is already super useful 😃

Thanks for asking, @vince-fugnitto. I am fine merging it, go ahead. I wanted to wait if there is something urgent we have to address before the merge. If something is wrong with the publishing job, I can take care of it today.

**Comment by kittaakos** [comment]
> If something is wrong with the publishing job, I can take care of it today.

Oh, it won't work. We do not have the secrets 😄

**Comment by vince-fugnitto** [comment]
@kittaakos I'll let you merge when you're comfortable :)

**Comment by marechal-p** [comment]
@kittaakos @vince-fugnitto is there something we need to wait for before merging?

**Comment by kittaakos** [comment]
> @kittaakos @vince-fugnitto is there something we need to wait for before merging?

> We do not have the secrets 😄

@marcdumais-work, if you're fine with it, I can push the secrets [this](https://github.com/eclipse-theia/theia/issues/8772#issuecomment-733079667) way.

**Comment by vince-fugnitto** [comment]
> > @kittaakos @vince-fugnitto is there something we need to wait for before merging?
> 
> > We do not have the secrets smile
> 
> @marcdumais-work, if you're fine with it, I can push the secrets [this](https://github.com/eclipse-theia/theia/issues/8772#issuecomment-733079667) way.

@kittaakos we discussed and it should be fine, you can merge the pull-request as is (with build and test) and the secrets can be set whenever available (it should not block the pull-request).

**Pull request merged**
